### ISSUE TITLE
Europe import gets stuck on ways

### ISSUE THREAD

**Issue by lhopki01** [osm-search/Nominatim#1978]
A full import of https://download.geofabrik.de/europe-latest.osm.pbf gets stuck as soon as it gets to ways.

```
Processing: Node(2605985k 456.3k/s) Way(638k 0.18k/s) Relation(0 0.0/s)
```

I have tried it with both v1.20 of osm2pgsql that comes with Nominatim v3.5.1 and also osm2pgsql 1.3.0 which is the most up to date version.

I have found similar issues such as:
https://github.com/osm-search/Nominatim/issues/1550  However that seems to be issues with osm2pgsql v1.21 
and https://github.com/openstreetmap/osm2pgsql/issues/534  However that appears to be an issue with disk speed and my `ha` is very low.

I'm running the import on a 96core machine with 50GB of RAM and 750GB SSD.  It has worked previously very quickly for other regions.  My `--osm2pgsql-cache` is set to 25000 which is enough to hold the entire pbf file in memory.

My postgres.conf is:
```
# Defaults from postgres db init
max_connections = 100			# (change requires restart)
dynamic_shared_memory_type = posix	# the default is the first option
min_wal_size = 80MB
log_timezone = 'Etc/UTC'
datestyle = 'iso, mdy'
timezone = 'Etc/UTC'
lc_messages = 'C.UTF-8'			# locale for system error message
lc_monetary = 'C.UTF-8'			# locale for monetary formatting
lc_numeric = 'C.UTF-8'			# locale for number formatting
lc_time = 'C.UTF-8'				# locale for time formatting

# Based on https://wiki.openstreetmap.org/wiki/PostgreSQL
shared_buffers = 4GB
maintenance_work_mem = 10GB
autovacuum_work_mem = 1GB
work_mem = 256MB
checkpoint_timeout = 10min
max_wal_size = 25GB
```

Output of top as well:
```
top - 03:21:56 up  7:36,  0 users,  load average: 1.10, 1.09, 1.04
Tasks:  19 total,   1 running,  18 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.1 us,  0.0 sy,  0.0 ni, 99.7 id,  0.2 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :  87092.1 total,  30475.4 free,  27253.4 used,  29363.3 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.  54874.0 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                                                                                              
     38 postgres  20   0 4404008   4.1g   4.1g S   1.7   4.8   1:30.25 postgres                                                                                                                                                                                             
    169 postgres  20   0 4392784   4.1g   4.1g D   1.3   4.8   4:41.63 postgres                                                                                                                                                                                             
    165 nominat+  20   0   27.2g  24.7g  11836 S   1.0  29.1  31:39.34 osm2pgsql                                                                                                                                                                                            
     39 postgres  20   0 4391716   4.0g   4.0g S   0.3   4.7   0:23.79 postgres                                                                                                                                                                                             
      1 root      20   0    8700   3572   3276 S   0.0   0.0   0:00.01 init.sh                                                                                                                                                                                              
     36 postgres  20   0 4391580 127672 125776 S   0.0   0.1   0:00.97 postgres                                                                                                                                                                                             
     40 postgres  20   0 4391580  21996  20080 S   0.0   0.0   1:31.52 postgres                                                                                                                                                                                             
     41 postgres  20   0 4392120   8544   6328 S   0.0   0.0   0:00.33 postgres                                                                                                                                                                                             
     42 postgres  20   0   71768   4808   2912 S   0.0   0.0   0:13.07 postgres                                                                                                                                                                                             
     43 postgres  20   0 4392004   6920   4828 S   0.0   0.0   0:00.00 postgres                                                                                                                                                                                             
     60 root      20   0   10080   3880   3520 S   0.0   0.0   0:00.00 sudo                                                                                                                                                                                                 
     61 root      20   0    7240    532    464 S   0.0   0.0   0:00.69 tee                                                                                                                                                                                                  
     62 nominat+  20   0   78824  25384  19972 S   0.0   0.0   0:00.03 setup.php                                                                                                                                                                                            
     67 postgres  20   0 4410852  36408  32900 S   0.0   0.0   0:00.01 postgres                                                                                                                                                                                             
    164 nominat+  20   0    2612    608    540 S   0.0   0.0   0:00.00 sh                                                                                                                                                                                                   
    168 postgres  20   0 4410044   4.1g   4.1g S   0.0   4.8  62:42.30 postgres                                                                                                                                                                                             
    171 postgres  20   0 4427500   1.9g   1.9g S   0.0   2.3   4:04.50 postgres                                                                                                                                                                                             
    224 root      20   0    9104   4196   3484 S   0.0   0.0   0:00.12 bash                                                                                                                                                                                                 
   1817 root      20   0   10944   3848   3204 R   0.0   0.0   0:00.06 top                                                                                                                                                                                                  
```

**Comment by lonvia** [comment]
What version of PostgreSQL and PostGIS do you use?

Have you configured a [flatnode file](https://nominatim.org/release-docs/latest/admin/Import-and-Update/#flatnode-files)? It would be strongly recommended for a Europe import.

**Comment by lhopki01** [comment]
>What version of PostgreSQL and PostGIS do you use?
```
postgresql-12-postgis-3-scripts/now 3.0.0+dfsg-6ubuntu4 all
postgresql-12-postgis-3/now 3.0.0+dfsg-6ubuntu4 amd64
postgresql-12/now 12.2-4 amd64
```
>Have you configured a flatnode file? It would be strongly recommended for a Europe import.
I haven't.  From reading I got the impression that it would save me disk space and RAM on import but would actually be slower to import so I've avoided it.

**Comment by lonvia** [comment]
> I haven't. From reading I got the impression that it would save me disk space and RAM on import but would actually be slower to import so I've avoided it.

On second thought, it would make import of nodes a lot faster but shouldn't have too much impact on the way imports when you have configured 25GB of cache.

Excuse the obvious question: have you restarted (and not only reloaded) PostgreSQL after changing its configuration?

**Comment by lonvia** [comment]
The other thing to consider: is this running in a VM? If so, can you check that the disks are not a bottleneck?

**Comment by lhopki01** [comment]
>Excuse the obvious question: have you restarted (and not only reloaded) PostgreSQL after changing its configuration?

Yes it was the first thing I checked when it started going slowly.  I logged into postgres and check all params that I set were correctly set.  I actually init a fresh db.  Copy the conf file in and then start the db.  I've included my script below.

>The other thing to consider: is this running in a VM? If so, can you check that the disks are not a bottleneck?

It is running on a VM.  Actually it's docker container within a VM.  I don't think the it's a disk bottleneck as top shows disk wait access never going above 2%.

```bash
#!/bin/bash

set -euo pipefail

mkdir -p /tmp/data

echo "Downloading $NOMINATIM_PBF_URL"
curl --silent -L "$NOMINATIM_PBF_URL" --create-dirs -o /tmp/data/nominatim-data.osm.pbf

if [ "${DOWNLOAD_GB_POSTCODES:-false}" = true ]; then
  echo "Downloading gb postcodes"
  curl --silent -L "https://www.nominatim.org/data/gb_postcode_data.sql.gz" -o /app/src/data/gb_postcode_data.sql.gz
fi
#RUN curl --silent -L "https://www.nominatim.org/data/us_postcode_data.sql.gz" -o /app/src/data/us_postcode_data.sql.gz

echo "Starting postgres"
mkdir -p "$PGDATA"
sudo chown -R postgres:postgres "$PGDATA"

sudo -u postgres /usr/lib/postgresql/12/bin/initdb -D "$PGDATA"
sudo cp /app/postgresql-init.conf "$PGDATA/postgresql.conf"
sudo -u postgres /usr/lib/postgresql/12/bin/pg_ctl -D "$PGDATA" start

echo "Creating users"
sudo -u postgres createuser -s nominatim
sudo -u postgres createuser -SDR www-data
useradd -m -p password1234 nominatim
chown -R nominatim:nominatim ./src

echo "Importing $NOMINATIM_PBF_URL"
sudo -u nominatim ./src/build/utils/setup.php --osm-file /tmp/data/nominatim-data.osm.pbf --all --osm2pgsql-cache "$OSM2PGSQL_CACHE" --threads "$THREADS" --drop 2>&1 | tee > /tmp/setup.log
sudo -u nominatim ./src/build/utils/check_import_finished.php
sudo -u postgres /usr/lib/postgresql/12/bin/pg_ctl -D "$PGDATA" stop
```

**Comment by lonvia** [comment]
Then I'm out of ideas. I'm not aware of any inherent problems with Nominatim. I've just successfully imported a planet in under two days in a non-VM but otherwise similar setup. So the problem is very likely somewhere in your configuration. Try out the recommended Postgresql configuration from https://nominatim.org/release-docs/latest/admin/Installation/#postgresql-tuning, in particual `synchronous_commit` and `fsync`. Try with a flatnode file. Check the system and postgres log files of your VM and the system log files of the host machine for any unusual messages.

**Comment by lhopki01** [comment]
So you were right.  It was disk related.  It wasn't manifesting as high wait times for disk access but rather just very few reads per second.  I suppose it's a lot of sequential requests so while the disk doesn't look overloaded it is still slowing things down.

Since fixing the disk issues the osm2pgsql import step now takes about 1000s for the France region.  However it still seems very slow.  On rank 26 I get an average processing of 4706.692 per second and on rank 30 I get an average processing of 85.136 per second meaning that a full import of the France region takes about 1.5 days. 

>Try out the recommended Postgresql configuration from https://nominatim.org/release-docs/latest/admin/Installation/#postgresql-tuning, in particual synchronous_commit and fsync.

I have followed that guide and through running multiple tests found some settings that works slightly faster for my setup.  Here is my `postgresql.conf` file.

```
# Defaults from postgres db init

dynamic_shared_memory_type = posix	# the default is the first option
min_wal_size = 80MB
log_timezone = 'Etc/UTC'
datestyle = 'iso, mdy'
timezone = 'Etc/UTC'
lc_messages = 'C.UTF-8'			# locale for system error message
lc_monetary = 'C.UTF-8'			# locale for monetary formatting
lc_numeric = 'C.UTF-8'			# locale for number formatting
lc_time = 'C.UTF-8'				# locale for time formatting

# Based on https://nominatim.org/release-docs/latest/admin/Installation/#postgresql-tuning

shared_buffers = 8GB
maintenance_work_mem = 20GB
autovacuum_work_mem = 2GB
work_mem = 256MB
effective_cache_size = 24GB
synchronous_commit = off
max_wal_size = 10GB
checkpoint_timeout = 10min
checkpoint_completion_target = 0.9

fsync = off
full_page_writes = off

max_connections = 150
```

My import commands is:
```
sudo -u nominatim ./src/build/utils/setup.php --osm-file /tmp/data/nominatim-data.osm.pbf --all --osm2pgsql-cache 250000 --threads 80 --drop 2>&1 | tee /tmp/setup.log
```

This is running on a machine with 150GB of ram and 96 cores.  What I haven't found is any guidance on the number of threads to use.  Through multiple trials I've found that 80 seems to be fastest.

All this is to say I still don't understand why everything is going so slowly.

**Comment by lonvia** [comment]
Can you please try adding the following setting to your postgresql conf:

` ` ` 
random_page_cost = 1.5
cpu_tuple_cost = 0.01
` ` ` 

If you currently have the import running and it is busy indexing at rank 30, you can simply reload postgresql and observe if indexing gets faster. As a rule of thumb rank 30 should process at about the same rate as rank 26.

**Comment by lhopki01** [comment]
I set those as suggested above and am still getting 
```
4786.996 per second - rank 26
77.965 per second - rank 30
```
That being said `cpu_tuple_cost` is 0.01 by default anyway right?

**Comment by lhopki01** [comment]
So I think I have figured out what was causing the issue.  Earlier when I had the slow PBF parser step I had tried updating osm2pgsql by running the command below before compilation.
```
    git submodule update --remote osm2pgsql
```
I removed the line and now get `4876.054 per second - rank 26` and `11548.700 per second - rank 30`.

I'm not sure why an updated version of osm2pgsql would affect the import speed of rank 30.  Is this known or is this a landmine waiting to happen when it's updated in future?
